### ISSUE TITLE
Add HTTP CORS support

### ISSUE THREAD

**Issue by jprante** [line/armeria#205]
I have a hard time to understand why CorsHandler does not work with Armeria's ServerInitializer.

I tried this in `ServerInitializer`:

```
    private void configureHttp(ChannelPipeline p) {
        p.addLast(new Http2PrefaceOrHttpHandler());
        p.addLast(new CorsHandler(CorsConfigBuilder.forAnyOrigin().allowedRequestMethods(HttpMethod.POST).build()));
        configureRequestCountingHandlers(p);
        p.addLast(new HttpServerHandler(config, SessionProtocol.H1C));
    }

```

with the test code

```
    @Test
    public void testCORSPreflight() throws Exception {
        SimpleHttpClient client = Clients.newClient(remoteInvokerFactory, "none+http://127.0.0.1:" + httpPort,
                SimpleHttpClient.class);
        // pre-flight request
        SimpleHttpRequest request = SimpleHttpRequestBuilder.forOptions("/cors")
                .header(HttpHeaderNames.ACCEPT, "utf-8")
                .header(HttpHeaderNames.ORIGIN, "http://example.com")
                .header(HttpHeaderNames.ACCESS_CONTROL_REQUEST_METHOD, HttpMethod.POST.asciiName())
                .build();
        SimpleHttpResponse response = client.execute(request).get();
        assertEquals(HttpResponseStatus.OK, response.status());
        assertEquals("http://example.com", response.headers().get(HttpHeaderNames.ACCESS_CONTROL_ALLOW_ORIGIN));
    }
```

but it fails with NPE

```
00:36:58.441 [armeria-server-nio-3-211] WARN  c.l.armeria.server.HttpServerHandler - [id: 0x89f474cc, L:/127.0.0.1:52981 - R:/127.0.0.1:52985][h2c] Unexpected exception:
java.lang.NullPointerException: null
	at io.netty.handler.codec.http.cors.CorsHandler.setOrigin(CorsHandler.java:97)
	at io.netty.handler.codec.http.cors.CorsHandler.write(CorsHandler.java:197)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:748)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:740)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:826)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:733)
	at io.netty.handler.timeout.IdleStateHandler.write(IdleStateHandler.java:284)
	at com.linecorp.armeria.server.HttpServerIdleTimeoutHandler.write(HttpServerIdleTimeoutHandler.java:75)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:748)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:740)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:826)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:733)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:714)
	at com.linecorp.armeria.server.HttpServerHandler.respond(HttpServerHandler.java:556)
	at com.linecorp.armeria.server.HttpServerHandler.respond(HttpServerHandler.java:524)
	at com.linecorp.armeria.server.HttpServerHandler.respond(HttpServerHandler.java:497)
	at com.linecorp.armeria.server.HttpServerHandler.handleNonExistentMapping(HttpServerHandler.java:322)
	at com.linecorp.armeria.server.HttpServerHandler.handleRequest(HttpServerHandler.java:244)
	at com.linecorp.armeria.server.HttpServerHandler.channelRead(HttpServerHandler.java:190)
	at com.linecorp.armeria.server.HttpServerHandler.userEventTriggered(HttpServerHandler.java:652)
```

I'm not very familiar with Armeria, I wonder why `channelRead` in `CorsHandler` is not called any more. The `write` method is called, but it assumes there is always a non-null request set.

Maybe the solution is trivial, maybe a different `CorsHandler` must be implemented?

**Comment by trustin** [comment]
`ServerInitializer` and its related classes manipulate the pipeline dynamically. There are a few reasons I can think of:

- `CorsHandler` has been placed in a wrong place eventually due to pipeline manipulation.
- Given that you used `SimpleHttpClient`, it is very likely that the connection has been upgraded to HTTP/2. As far as I know, `CorsHandler` does not work with an HTTP/2 connection yet.

Actually, it seems like a good idea to write a CORS handler as an Armeria service decorator, so that it works for both HTTP/1 and 2 and a user does not need to modify Armeria source code by him/herself. What do you think? Would you be interested in filing an issue or even a PR?

Another question: what do you think about Armeria? What could be done better?

**Comment by trustin** [comment]
Fixed in 0.26.0. Thanks @jprante to his initial work. 👍
